Added Chests of Agony

hannahbee91 opened the pull request:
This feature makes it so that the "Chest Size and Texture Matches Contents" takes effect when the player has the Stone of Agony. This is disabled by default and the checkbox only appears if the player selects anything other than "Disabled" for the Chest Size enhancement.

This suggestion was brought up in a ZFG1 rando steam on 01/11/2022 so I decided to take a crack at it. It was easier than I thought.

Commit:
Added a feature so that chests only change size/texture if the player has the stone of agony

Commit:
Cleaned up a line that I left in from bad attempts

hannahbee91 commented:
https://user-images.githubusercontent.com/20977880/199392954-0d166b3a-6945-49dc-9545-cdac5dc417a8.mp4

Video showing that these changes work, demonstrating how changing settings and quest item status affects chest display.

stratomaster64 commented:
Just one question: does having Chests of Agony also affect the collision of the chests if you change them midgame like you did in your video? Small and large chests have different collision (which can lead to some tricks in glitched/no logic) when using CSMC settings normally.

hannahbee91 commented:
Yes, it does affect the collision. Toggling it on/off while pushed right up against a small chest will push link out and away when the chest enlarges.

Even changing size during an opening animation, it moves around Link's position if collision changes.


https://user-images.githubusercontent.com/20977880/199395693-08eb52f1-3b7c-428f-bfb3-6150014efd2e.mp4

stratomaster64 commented:
Okie dokie, so it works as expected. Honestly, I think this would be better if chests of agony were set on seed generation since having the ability toggleable at will might break logic with certain chests in certain sizes, but I'll leave that for the devs to comment on.

garrettjoecox commented on soh/soh/GameMenuBar.cpp:
```diff
@@ -811,6 +813,12 @@ namespace GameMenuBar {
+                        UIWidgets::PaddedEnhancementCheckbox("Chests of Agony", "gChestSizeDependsStoneOfAgony", true, false);
+                        UIWidgets::Tooltip("Only change the size/texture of chests if you have the Stone of Agony.");
+                    } else {
+                        CVar_SetS32("gChestSizeDependsStoneOfAgony", 0);
```
Currently this line will run every frame, instead we need to do an `if` around the ` UIWidgets::EnhancementCombobox("gChestSizeAndTextureMatchesContents"` above, which will only update this when that value is changed

hannahbee91 replied:
From my understanding of all of the other conditional options in `GameMenuBar.cpp`, `if`s can only be around the `ImGui::BeginMenu()` classes, not individual `UIWidgets` objects. I was using examples from options such as  [the checkbox for Kokiri Draw Distance](https://github.com/da-code-a/Shipwright/blob/9b76c756f56bb104185257758371e0c912e88932/soh/soh/GameMenuBar.cpp#L1314-L1319) and the options concerning [hammer crouch-stab](https://github.com/da-code-a/Shipwright/blob/9b76c756f56bb104185257758371e0c912e88932/soh/soh/GameMenuBar.cpp#L1150-L1155). All other examples of `UIWidgets` objects appearing after a value has been set elsewhere follow this same structure.

garrettjoecox replied:
All interactable ImGui elements should return a bool that returns true on the frame that the input was interacted with. I've updated some of our UIWidgets to expose this as it should be, but it looks like `EnhancementCombobox` and a few others have not exposed this yet.

Anywhere else this pattern appears (I see it is indeed like this for Kokiri Draw Distance) is a not desired, and should also be updated to only be set on user interaction.

garrettjoecox replied:
I can push up a PR fixing the kokiri draw distance instance and exposing the bool via the combobox, or you can add the exposing the bool in this PR, and I can submit a follow up fixing the kokiri draw distance

PurpleHato replied:
> I can push up a PR fixing the kokiri draw distance instance and exposing the bool via the combobox, or you can add the exposing the bool in this PR, and I can submit a follow up fixing the kokiri draw distance

My bad for this one, didn't knew doing this would be an every frame check

hannahbee91 replied:
I went ahead and fixed Kokiri Draw as well as exposing the bool via combobox in 3e44805d.

Let it be known, I'm not a C++ dev; I work mostly with Python. So, if there are any other conventions or things I'm missing, let me know and I'll be happy to improve :)

Commit:
Making the Kokiri draw distance check not run every frame

Commit:
Made Chests of Agony option not run every frame

Also enabled all `EnhancementCombobox` elements to be used as bools

garrettjoecox commented on soh/soh/UIWidgets.cpp:
```diff
@@ -227,7 +227,7 @@ namespace UIWidgets {
     }
 
-    void EnhancementCombobox(const char* name, const char* ComboArray[], size_t arraySize, uint8_t FirstTimeValue) {
+    bool EnhancementCombobox(const char* name, const char* ComboArray[], size_t arraySize, uint8_t FirstTimeValue) {
```
You'll still need to make the change to the method definition to actually return the bool here, you can reference `EnhancementCheckbox` for an example

garrettjoecox commented on soh/soh/GameMenuBar.cpp:
```diff
@@ -801,7 +803,14 @@ namespace GameMenuBar {
                     const char* chestSizeAndTextureMatchesContentsOptions[4] = { "Disabled", "Both", "Texture Only", "Size Only"};
-                    UIWidgets::EnhancementCombobox("gChestSizeAndTextureMatchesContents", chestSizeAndTextureMatchesContentsOptions, 4, 0);
+                    if(UIWidgets::EnhancementCombobox("gChestSizeAndTextureMatchesContents", chestSizeAndTextureMatchesContentsOptions, 4, 0)) {
+                        if (CVar_GetS32("gChestSizeAndTextureMatchesContents", 0) > 0) {
```
So the first part of this condition should still be outside of the condition you added, otherwise the additional "Chests of Agony" checkbox will only render the frame you interact with the above combobox. (The same goes for the kokiri distance change)

Commit:
Making it so that the additional Kokiri Draw and Chests of Agony checkboxes appear for more than a single frame.

The check to disable these child settings when their parent settings are turned off still only runs on the frame that the parents are toggled.

garrettjoecox commented on soh/soh/GameMenuBar.cpp:
```diff
@@ -801,7 +803,15 @@ namespace GameMenuBar {
+                        UIWidgets::PaddedEnhancementCheckbox("Chests of Agony", "gChestSizeDependsStoneOfAgony", true, false);
+                        UIWidgets::Tooltip("Only change the size/texture of chests if you have the Stone of Agony.");
+                    }
                     UIWidgets::Tooltip(
```
This tooltip should move up one block, currently `gChestSizeAndTextureMatchesContents` has no tooltip, and `gChestSizeDependsStoneOfAgony`'s tooltip is replaced with this one

garrettjoecox commented on soh/soh/UIWidgets.cpp:
```diff
@@ -245,7 +246,9 @@ namespace UIWidgets {
                 }
             }
             ImGui::EndCombo();
+            changed = true;
```
This should move up to line 243, within the `if (ImGui::Selectable...`

garrettjoecox reviewed (commented):
Sorry I know I'm throwing you for a loop on this one, just a few more tweaks 😅

Commit:
Fix tool tip and boolean switch

hannahbee91 commented:
> Sorry I know I'm throwing you for a loop on this one, just a few more tweaks :sweat_smile:

No worries :sweat_smile: . I addressed those couple changes, verified that the tooltip appears where it should again:

![image](https://user-images.githubusercontent.com/20977880/199566146-3342369b-5a7b-4ff0-80f8-29ae2ed2cc4d.png)

garrettjoecox commented on soh/src/overlays/actors/ovl_En_Box/z_en_box.c:
```diff
@@ -631,7 +634,7 @@ void EnBox_UpdateSizeAndTexture(EnBox* this, GlobalContext* globalCtx) {
     }
 
-    if (globalCtx->sceneNum != SCENE_TAKARAYA && (cvar == 1 || cvar == 3)) {
+    if (globalCtx->sceneNum != SCENE_TAKARAYA && (cvar == 1 || cvar == 3) && ((agonyCVar > 0 && stoneCheck) | agonyCVar == 0)) {
```
Looks like you missed one spot this condition needs to be added, down below at 665

hannahbee91 replied:
No I didn't 😜

garrettjoecox reviewed (approved):
Aside from that last comment this all LGTM, thanks for bearing with me through the review 😅

Commit:
Missed one condition

briaguya0 commented on soh/soh/UIWidgets.hpp:
```diff
@@ -36,7 +36,7 @@ namespace UIWidgets {
     bool PaddedEnhancementCheckbox(const char* text, const char* cvarName, bool padTop = true, bool padBottom = true, bool disabled = false, const char* disabledTooltipText = "", CheckboxGraphics disabledGraphic = CheckboxGraphics::Cross);
     void EnhancementCombo(const std::string& name, const char* cvarName, const std::vector<std::string>& items, int defaultValue = 0);
-    void EnhancementCombobox(const char* name, const char* ComboArray[], size_t arraySize, uint8_t FirstTimeValue);
+    bool EnhancementCombobox(const char* name, const char* ComboArray[], size_t arraySize, uint8_t FirstTimeValue);
```
i feel like having this be a bool is kind of odd, might make more sense to have it return an int because that's what the value in it is

garrettjoecox replied:
@briaguya-ai see the top resolved comment above, all of our wrapper ImGui components should be returning bools, consistent with ImGui and how it's supposed to be used

garrettjoecox commented:
@da-code-a Looks like I created some conflicts that will need to be resolved

Commit:
Merge branch 'develop' into chests_only_change_with_agony

hannahbee91 commented:
> @da-code-a Looks like I created some conflicts that will need to be resolved

Done!

dcvz commented:
@da-code-a im ready to merge this.. but looks like there's conflicts again. Probably due to #1927 - as soon as that's fixed I'll merge this

Commit:
Merge branch 'develop' into chests_only_change_with_agony

hannahbee91 commented:
@dcvz Great to hear, thank you! I've resolved those conflicts once again for you.

dcvz commented:
Good work here @da-code-a
